references/depth/stereo/cascade_evaluation.py

```diff
@@ -0,0 +1,242 @@
+    parser.add_argument("--num-workers", type=int, default=0, help="number of workers")
+    
+    parser.add_argument("--resize-size", type=int, n_args='+', default=[384, 512], help="resize size")
+    parser.add_argument("--n_iterations", nargs='+', type=int, default=[10], help="number of recurent iterations")
+    parser.add_argument("--n_cascades", nargs='+', type=int, default=[1], help="number of cascades")
+    parser.add_argument("--device", type=str, default="cuda", help="device to use")
+    
+    parser.add_argument("--save-predictions", action="store_true", help="save predictions")
+    
+    return parser
+
+def cascade_inference(model, image_left, image_right, iterations, cascades):
+    # check that image size is divisible by 16 * (2 ** (cascades - 1))
+    for image in [image_left, image_right]:
+        if image.shape[-2] % ((2 ** (cascades - 1))) != 0:
+            raise ValueError(f"image height is not divisible by {16 * (2 ** (cascades - 1))}. Image shape: {image.shape[-2]}")
+    
+        if image.shape[-1] % ((2 ** (cascades - 1))) != 0:
+            raise ValueError(f"image width is not divisible by {16 * (2 ** (cascades - 1))}. Image shape: {image.shape[-2]}")
+    
+    left_image_pyramid = [image_left]
+    right_image_pyramid = [image_right]
+    for idx in range(0, cascades - 1):
+        ds_factor = int(2 ** (idx + 1))
+        ds_shape = (image_left.shape[-2] // ds_factor, image_left.shape[-1] // ds_factor)
+        left_image_pyramid += F.interpolate(image_left, size=ds_shape, mode="bilinear", align_corners=True).unsqueeze(0)
+        right_image_pyramid += F.interpolate(image_right, size=ds_shape, mode="bilinear", align_corners=True).unsqueeze(0)
+        
+    flow_init = None
+    for left_image, right_image in zip(reversed(left_image_pyramid), reversed(right_image_pyramid)):
+        flow_pred = model(left_image, right_image, flow_init, iterations=iterations)
+        # flow pred is a list
+        flow_init = flow_pred[-1]
+        
+    return flow_init
+
+@torch.inference_mode()
+def _evaluate(
+    model, args, val_dataset, *, padder_mode, print_freq=10, writter=None, step=None, iterations=10, cascades=1, batch_size=None, header=None, save_images=False, save_path="",
+):
+    """Helper function to compute various metrics (epe, etc.) for a model on a given dataset.
+    We process as many samples as possible with ddp.
+    """
+    model.eval()
+    header = header or "Test:"
+    device = torch.device(args.device)
+    metric_logger = utils.MetricLogger(delimiter="  ")
+    
+    effective_batch_size = batch_size * args.world_size
+    can_do_distributed_test = effective_batch_size < len(val_dataset)
+    
+    if args.distributed and can_do_distributed_test:
+        sampler = torch.utils.data.distributed.DistributedSampler(val_dataset, shuffle=False, drop_last=True)
+    else:
+        sampler = torch.utils.data.SequentialSampler(val_dataset)
+    
+    val_loader = torch.utils.data.DataLoader(
```

Comment:
Why do we create `val_loader` inside `_evaluate` functions? Is there any reason we choose to do this instead of passing `val_loader`?

(same question in train.py)

Reply:
Because we might want to consume the specific dataset multiple times in the cascaded inference, therefore that's done to avoid any bugs where the dataloader might not be fully consumed and just be partially iterated over.

Reply:
Lets discuss about this tomorrow, I didn't really get what you mean here